libcontainerd/supervisor/remote_daemon.go

```diff
@@ -276,26 +282,28 @@ func (r *remote) monitorDaemon(ctx context.Context) {
 			}
 		}
 
-		tctx, cancel := context.WithTimeout(ctx, healthCheckTimeout)
-		_, err := client.IsServing(tctx)
-		cancel()
-		if err == nil {
-			if !started {
-				close(r.daemonStartCh)
-				started = true
+		if client != nil {
```

Comment:
Maybe do this instead:

```go
               if r.daemonPid == -1 {
                        <.....>
                }

                if client == nil {
                        client, err = containerd.New(r.GRPC.Address, containerd.WithTimeout(60*time.Second))
                        if err != nil {
                                r.logger.WithError(err).Error("failed connecting to containerd")
                                delay = time.After(100 * time.Millisecond)
                                continue
                        }
                }

                tctx, cancel := context.WithTimeout(ctx, healthCheckTimeout)
                _, err := client.IsServing(tctx)
                cancel()
....
```

Reply:
This appears to be simpler and easier to follow/read.

Reply:
I originally had a change like this, the problem I was concerned about though is if the daemon must be restarted in order to get a successful connection. There is no timeout in this flow to kill the daemon process. My thinking with the current change is that the connection logic may have already waited up to a reasonable time to connect, and if it couldn't then it should restart the whole flow, including killing containerd and starting a new process.

Reply:
makes sense